src/buf_writer.rs

```diff
@@ -0,0 +1,175 @@
+use {AsyncWrite, DEFAULT_BUF_SIZE};
+
+use bytes::{Buf, BufMut, BytesMut};
+use futures::{Async, Poll};
+
+use std::{cmp, fmt};
+use std::io::{self, SeekFrom};
+
+/// Wraps a writer and buffers its output.
+pub struct BufWriter<W> {
+    inner: W,
+    buf: io::Cursor<BytesMut>,
+}
+
+impl<W: io::Write> BufWriter<W> {
+    /// Creates a new `BufWriter` with a default buffer capacity.
+    pub fn new(inner: W) -> BufWriter<W> {
+        BufWriter::with_capacity(DEFAULT_BUF_SIZE, inner)
+    }
+
+    /// Creates a new `BufWriter` with the specified buffer capacity.
+    pub fn with_capacity(cap: usize, inner: W) -> BufWriter<W> {
+        BufWriter {
+            inner: inner,
+            buf: io::Cursor::new(BytesMut::with_capacity(cap)),
+        }
+    }
+
+    fn flush_once(&mut self) -> io::Result<()> {
+        if !self.buf.has_remaining() {
+            return Ok(());
+        }
+
+        self.do_flush()
+    }
+
+    fn flush_all(&mut self) -> io::Result<()> {
+        while self.buf.has_remaining() {
+            try!(self.do_flush());
+        }
+
+        Ok(())
+    }
+
+    fn do_flush(&mut self) -> io::Result<()> {
+        debug_assert!(self.buf.has_remaining());
+
+        match try!(self.inner.write(self.buf.bytes())) {
+            0 => {
+                return Err(io::Error::new(io::ErrorKind::WriteZero, "failed to
+                                          write the buffered data"));
+            }
+            n => self.buf.advance(n),
+        }
+
+        self.compact_buf();
+
+        Ok(())
+    }
+
+    fn compact_buf(&mut self) {
+        if self.buf.position() as usize == self.buf.get_ref().len() {
+            // Fully written, clear the buffer
+            self.buf.set_position(0);
+            self.buf.get_mut().clear();
+        }
+    }
+}
+
+impl<W> BufWriter<W> {
+    /// Gets a reference to the underlying writer.
+    pub fn get_ref(&self) -> &W {
+        &self.inner
+    }
+
+    /// Gets a mutable reference to the underlying writer.
+    ///
+    /// It is inadvisable to directly write to the underlying writer.
+    pub fn get_mut(&mut self) -> &mut W {
+        &mut self.inner
+    }
+
+    /// Unwraps this `BufWriter`, returning the underlying writer.
+    ///
+    /// The buffer is written out before returning the writer.
+    pub fn into_inner(self) -> W {
+        self.inner
+    }
+}
+
+impl<W: io::Write> io::Write for BufWriter<W> {
```

Comment:
I think one way to simplify this perhaps would be require `W: AsyncWrite` and just delegate to `AsyncWrite::write_buf` below perhaps?